GoogleCloudPlatform/bank-of-anthos
# IaC: Terraform and ACM configurations for Multi-Autopilot GKE environment setup with CI/CD, Anthos & CloudSQL

### Background 
Initially the fork that this PR is based on was created to showcase Google Cloud's CICD story encompassing Cloud Build / Cloud Deploy / kustomize / skaffold. In order to maximize value from the work done for the showcase we decided to automate the setup with terraform utilising Cloud Foundation Toolkit.

### Change Summary
Added configuration that includes:
- Terraform for all infrastructure (with exception of Github repo connection which is a one-time manual step)
- CI per team
  - CloudBuild /w skaffold (includes caching)
- CD (staging, production) per team
  - CloudDeploy /w skaffold profiles per environment
- 3 Environments:
  - Development (skaffold dev only)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload identity
    - Namespace: bank-of-anthos-development
  - Staging (CD deploy)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload identity
    - Namespace: bank-of-anthos-staging
    - CloudSQL
  - Production (CD deploy)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload Identity
    - Namespace: bank-of-anthos-production
    - CloudSQL
- Heavy use of kustomize components to stay DRY
- Minimized service account permissions
- Uses CloudFoundation Toolkit for Kubernetes (which restricts modularization of environments due to kubernetes provider circular dependency)

### Additional Notes
This PR does NOT include the changes to the src/ folder, which include kustomizations for microservices, skaffold profiles and cloudbuild.yaml definitions for individual teams. 
The terraform code included in this PR does however reference these files, which leads to this IaC not being immediately compatible with the current application source.

### Testing Procedure
Follow steps [here](https://github.com/aablsk/bank-of-anthos/blob/main/iac/tf-multienv-cicd-anthos-autopilot/README.md#how-to-set-it-up) using the prepared shell scripts for easiest setup.

Please note that script 4 and 5 will log errors as they require the second PR (which includes changes to src/ folder)

### Related PRs or Issues 
I stole shamelessly from these PRs:
https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/881
https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/792

Changes to src/ can be seen in this [fork](https://github.com/aablsk/bank-of-anthos)
